File: Que Hacer MDQ v2/Que Hacer MDQ v2/Classes/Activity.m
//
//  Activity.m
//  Que Hacer MDQ v2
//
//  Created by Lucas on 5/22/15.
//  Copyright (c) 2015 Globant iOS MDQ. All rights reserved.
//

#import "Activity.h"
#import "Schedule.h"


@implementation Activity

@dynamic areaId;
@dynamic contactPhone1;
@dynamic contactPhone2;
@dynamic contactPhone3;
@dynamic cost;
@dynamic costNumber;
@dynamic costString;
@dynamic desc;
@dynamic end;
@dynamic handicapAccessRamp;
@dynamic handicapRestroom;
@dynamic handicapRestroomInGroundFloor;
@dynamic activityId;
@dynamic locationDetails;
@dynamic locationHouseNumberingOrKm;
@dynamic locationLatitude;
@dynamic locationLongitude;
@dynamic locationStreetOrRoute;
@dynamic name;
@dynamic ocurrsOnce;
@dynamic paidParkingZone;
@dynamic photoUrl1;
@dynamic photoUrl2;
@dynamic photoUrl3;
@dynamic sharingUrl;
@dynamic sourceWebServiceName;
@dynamic start;
@dynamic visible;
@dynamic visitingHoursString;
@dynamic website;
@dynamic tags;
@dynamic schedules;


+(Activity*)persistentInstanceFromDictionary:(NSDictionary*)aDictionary
{
    NSManagedObjectContext* context = [[CoreDataHelper sharedInstance] managedObjectContext];
    Activity* newActivity = [NSEntityDescription insertNewObjectForEntityForName:@"Activity" inManagedObjectContext:context];
    [newActivity setFromDictionary:aDictionary];
    return newActivity;
}

-(Activity*)setFromDictionary:(NSDictionary*)aDictionary
{
    
    self.activityId = [DataTypesHelper stringToNSNumber:[aDictionary objectForKey:@"activityId"]];
    self.name = [aDictionary objectForKey:@"name"];
    self.desc = [aDictionary objectForKey:@"description"];
    self.contactPhone1 = [aDictionary objectForKey:@"contactPhone1"];
    self.contactPhone2 = [aDictionary objectForKey:@"contactPhone2"];
    self.contactPhone3 = [aDictionary objectForKey:@"contactPhone3"];
    self.photoUrl1 = [aDictionary objectForKey:@"photoUrl1"];
    self.photoUrl2 = [aDictionary objectForKey:@"photoUrl2"];
    self.photoUrl3 = [aDictionary objectForKey:@"photoUrl3"];
    self.locationStreetOrRoute = [aDictionary objectForKey:@"locationStreetOrRoute"];
    self.locationHouseNumberingOrKm = [aDictionary objectForKey:@"locationHouseNumberingOrKm"];
    self.locationDetails = [aDictionary objectForKey:@"locationDetails"];
    self.locationLatitude = [DataTypesHelper stringToNSNumber:[aDictionary objectForKey:@"locationLatitude"]];
    self.locationLongitude = [DataTypesHelper stringToNSNumber:[aDictionary objectForKey:@"locationLongitude"]];
    self.website = [aDictionary objectForKey:@"website"];
    self.visitingHoursString = [aDictionary objectForKey:@"visitingHoursString"];
    self.start = [DataTypesHelper qhmdqDateTimeStringToNSDate:[aDictionary objectForKey:@"start"]];
    self.end = [DataTypesHelper qhmdqDateTimeStringToNSDate:[aDictionary objectForKey:@"end"]];
    self.cost = [DataTypesHelper stringToNSNumber:[aDictionary objectForKey:@"cost"]];
    self.costString = [aDictionary objectForKey:@"costString"];
    self.handicapAccessRamp = [DataTypesHelper stringToNSNumber:[aDictionary objectForKey:@"handicapAccessRamp"] ];
    self.handicapRestroom = [DataTypesHelper stringToNSNumber:[aDictionary objectForKey:@"handicapRestroom"]];
    self.handicapRestroomInGroundFloor = [DataTypesHelper stringToNSNumber:[aDictionary objectForKey:@"handicapRestroomInGroundFloor"]];
    self.paidParkingZone = [DataTypesHelper stringToNSNumber:[aDictionary objectForKey:@"paidParkingZone"]];
    self.sharingUrl = [aDictionary objectForKey:@"sharingUrl"];
    
    
    /*
    //Relationship: Tags
    NSArray* tagsSource = [aDictionary objectForKey:@"tags"];
    NSMutableArray* tagsArray = [[NSMutableArray alloc] init];
    for ( NSDictionary* tagDictionary in tagsSource) {
        [tagsArray addObject:[Tag instanceFromDictionary:tagDictionary]];
    }
    self.tags = [[NSSet alloc] initWithArray:tagsArray];
    */
    
    
    //Relationship: Tags
    NSMutableArray* tempMutableArray = [[NSMutableArray alloc] init];
    NSArray* tagsSource = [aDictionary objectForKey:@"tags"];
    //If the activity's JSON came with tags in it, associate them with the Tag objects
    if ([tagsSource count] > 0)
    {
        for ( NSDictionary* tagDictionary in tagsSource) {
            
            NSString* tagIdNSString = [tagDictionary objectForKey:@"tagId"];
            NSNumber* tagIdNSNumber = [DataTypesHelper stringToNSNumber:tagIdNSString];
            [tempMutableArray addObject:tagIdNSNumber];
            //NSString* tagIdString = [tagDictionary objectForKey:@"tagId"];
            //[tempMutableArray addObject:[tagDictionary objectForKey:@"tagId"]];
            //[newActivity addTagsObject:[[CoreDataHelper sharedInstance] getTagWithId:[tagDictionary objectForKey:@"tagId"]]] ;
        }
    }
    self.tags = [NSArray arrayWithArray:tempMutableArray];
    
    /*
    //Relationship: Schedules
    NSArray* schedulesSource = [aDictionary objectForKey:@"schedules"];
    NSMutableArray* schedulesArray = [[NSMutableArray alloc] init];
    for (NSDictionary* scheduleDictionary in schedulesSource)
    {
        [schedulesArray addObject:[Schedule instanceFromDictionary:scheduleDictionary]];
    }
    */
    
    //Relationship: Schedules
    NSManagedObjectContext* context = [[CoreDataHelper sharedInstance] managedObjectContext];
    NSArray* schedulesSource = [aDictionary objectForKey:@"schedules"];
    for (NSDictionary* scheduleDictionary in schedulesSource)
    {
        [self addSchedulesObject:[[NSEntityDescription insertNewObjectForEntityForName:@"Schedule" inManagedObjectContext:context] initWithDictionary:scheduleDictionary]];
    }
    
    NSError* error;
    [context save:&error];
    if (error)
    {
        NSLog(@"There was an error when trying to save the activity with id %@", self.activityId);
        return nil;
    }
    
    return self;
}



@end

File: Que Hacer MDQ v2/Que Hacer MDQ v2/Classes/DataManager/Activity.swift
//
//  Activity.swift
//  Que Hacer? MdQ
//
//  Created by Federico Gonzalez on 2/9/15.
//  Copyright (c) 2015 Globant iOS MDQ. All rights reserved.
//

import Foundation
import CoreData

@objc(Activity)
class Activity: NSManagedObject {
    
    @NSManaged var id: Int
    
    //Relationships:
    @NSManaged var feeds: NSSet
    @NSManaged var area: Area
    @NSManaged var tags: NSSet
    @NSManaged var schedules: NSSet

    //Atributes
    @NSManaged var locationLatitude: Float
    @NSManaged var locationLongitude: Float
    @NSManaged var locationStreetOrRoute: String
    @NSManaged var locationHouseNumberingOrKm: String
    @NSManaged var locationDetails: String  //in case of no specific data of street and number
    @NSManaged var contactPhone1: String
    @NSManaged var contactPhone2: String
    @NSManaged var contactPhone3: String
    @NSManaged var photoUrl1: String
    @NSManaged var photoUrl2: String
    @NSManaged var photoUrl3: String
    @NSManaged var name: String
    @NSManaged var desc: String
    @NSManaged var website: String //NSURL
    @NSManaged var usesSchedule: Bool
    @NSManaged var visitingHours: String
    @NSManaged var start: NSDate
    @NSManaged var end: NSDate
    @NSManaged var cost: Float
    @NSManaged var costString: String //In case of no numeric cost specified
    @NSManaged var handicapAccessRamp: Bool
    @NSManaged var handicapRestroom: Bool
    @NSManaged var paidParkingZone: Bool
    @NSManaged var sharingUrl: String
    @NSManaged var visible: Bool
    @NSManaged var createdAt: NSDate
    @NSManaged var updatedAt: NSDate
    
    
    //Inferred by other attributes for simplicity
//    @NSManaged var usesDateString: Bool
//    @NSManaged var usesTimeString: Bool
    
    
    
    //Returns true if the activity has more than one tag
    func hasMultipleTags() -> Bool {
        
        if self.tags.count > 1 {
            return true
        } else {
            return false
        }
    
    }
    
    //If the activity has only one tag it returns it, otherwise returns nil (with console error).
    func getSingleTag() -> Tag? {
        
        if self.hasMultipleTags() {
            NSLog("Activitie with more than 1 tag, cannot retrieve single tag.")
            return nil
        } else {
            return self.tags.allObjects[0] as? Tag
        }
        
    }
    
    
//    func occurrsOnce() -> Bool {
//        
//        return false
//    }
    
    
    //returns true if delete all was succesfull
    class func deleteAllActivities() -> Bool {
        
        var context: NSManagedObjectContext = DataAccessHelper.sharedInstance().managedObjectContext
        let fetchRequest = NSFetchRequest(entityName: "Activity")
        
        fetchRequest.includesPropertyValues = false
        
        var error: NSError? = nil
        var matches: NSArray = context.executeFetchRequest(fetchRequest, error: &error)!
        
        if (error != nil) {
            NSLog("Error in fetching AllActivities for delete in method deleteAllActivities.");
            return false;
        }
        
        for activity in matches {
            context.deleteObject(activity as NSManagedObject)
        }
        
        var saveError: NSError? = nil
        if !context.save(&saveError) {
            
            NSLog("Error, couldn't delete: %@", saveError!.localizedDescription)
            context.rollback()
            
            return false
        }
        
        return true
    }
    
    //Activities With Predicate
    class func getAllActivities(predicate: NSPredicate!) -> NSArray? {
        
        var context: NSManagedObjectContext = DataAccessHelper.sharedInstance().managedObjectContext
        let fetchRequest = NSFetchRequest(entityName: "Activity")
        
        //fetchRequest.includesPropertyValues = false
        fetchRequest.predicate = predicate
        
        var error: NSError? = nil
        var matches: NSArray = context.executeFetchRequest(fetchRequest, error: &error)!
        
        if (error != nil) && (matches.count > 0) {
            return matches
        } else {
            NSLog("Error while retrieving activities: %@", error!.localizedDescription)
            return nil
        }
        
    }
    /*
    //Activities With Sort Descriptor
    class func getAllActivities(sortDescriptor: NSSortDescriptor?) -> NSArray? {
        //optional sortDescriptor
        
        var context: NSManagedObjectContext = DataAccessHelper.sharedInstance().managedObjectContext
        let fetchRequest = NSFetchRequest(entityName: "Activity")
        
        //fetchRequest.includesPropertyValues = false
        if sortDescriptor != nil {
            fetchRequest.sortDescriptors = [sortDescriptor!]
        }
        
        var error: NSError? = nil
        var matches: NSArray = context.executeFetchRequest(fetchRequest, error: &error)!
        
        if (error != nil) && (matches.count > 0) {
            return matches
        } else {
            NSLog("Error while retrieving activities: %@", error!.localizedDescription)
            return nil
        }
        
    }
    */
    func initWithDictionary(dic: NSDictionary) -> Activity {
        //NOT FINISHED
        
        var context: NSManagedObjectContext = DataAccessHelper.sharedInstance().managedObjectContext
        let newActivity = NSEntityDescription.insertNewObjectForEntityForName("Activity",
            inManagedObjectContext: context) as Activity
        
        newActivity.id = dic.objectForKey("activityid") as Int
        newActivity.name = dic.objectForKey("name") as String
        newActivity.desc = dic.objectForKey("description") as String
        
        newActivity.contactPhone1 = dic.objectForKey("contactPhone1") as String
        newActivity.contactPhone2 = dic.objectForKey("contactPhone2") as String
        newActivity.contactPhone3 = dic.objectForKey("contactPhone3") as String
        
        newActivity.photoUrl1 = dic.objectForKey("photoUrl1") as String
        newActivity.photoUrl2 = dic.objectForKey("photoUrl2") as String
        newActivity.photoUrl3 = dic.objectForKey("photoUrl3") as String
        
        newActivity.locationStreetOrRoute = dic.objectForKey("locationStreetOrRoute") as String
        newActivity.locationHouseNumberingOrKm = dic.objectForKey("locationHouseNumberingOrKm") as String
        newActivity.locationDetails = dic.objectForKey("locationDetails") as String
        newActivity.locationLatitude = dic.objectForKey("locationLatitude") as Float
        newActivity.locationLongitude = dic.objectForKey("locationLongitude") as Float
        
        newActivity.website = dic.objectForKey("website") as String
        newActivity.usesSchedule = dic.objectForKey("usesSchedule") as Bool
        newActivity.visitingHours = dic.objectForKey("visitingHours") as String
        
        newActivity.start = dic.objectForKey("start") as NSDate
        newActivity.end = dic.objectForKey("end") as NSDate
        
        newActivity.cost = dic.objectForKey("cost") as Float
        newActivity.costString = dic.objectForKey("costString") as String
        
        newActivity.handicapAccessRamp = dic.objectForKey("handicapAccessRamp") as Bool
        newActivity.handicapRestroom = dic.objectForKey("handicapRestroom") as Bool
        
        newActivity.paidParkingZone = dic.objectForKey("paidParkingZone") as Bool
        newActivity.sharingUrl = dic.objectForKey("sharingUrl") as String
        
        newActivity.visible = dic.objectForKey("visible") as Bool
        newActivity.createdAt = dic.objectForKey("createdAt") as NSDate
        newActivity.updatedAt = dic.objectForKey("updatedAt") as NSDate
        
        //Relationships:
        newActivity.area = Area.getArea(dic.objectForKey("areaId") as Int)!
        
        if let tags = dic["tags"] as? [[String:String]] {
            for tag in tags {
                
                var tagId: Int = tag["tagId"]!.toInt()!
                var auxTag: Tag = Tag.getTag(tagId)!
                newActivity.addObject(auxTag, forKey: "tags")
            }
        } else {
            NSLog("Error trying to parse the tags for the activity.")
        }
        
        if let schedules = dic["schedules"] as? [[String:String]] {
            for dicSchedule in schedules {
                
                let newSchedule = NSEntityDescription.insertNewObjectForEntityForName("Schedule",
                                                    inManagedObjectContext: context) as Schedule
                
                newSchedule.startDate = NSDate(dateString: dicSchedule["dateStringStart"]!)
                newSchedule.endDate = NSDate(dateString: dicSchedule["dateStringEnd"]!)
                //newSchedule.startTime = NSDate(dateString: dicSchedule["timeStringStart"]!)
                //newSchedule.endTime = NSDate(dateString: dicSchedule["timeStringStart"]!)
                
                newActivity.addObject(newSchedule, forKey: "schedules")
            }
        } else {
            NSLog("Error trying to parse the schedule for the activity.")
        }
        
        return newActivity
    }
    

}




